### ISSUE TITLE
services/horizon: /order_book slower when reading from offers graph

### ISSUE THREAD

**Issue by bartekn** [stellar/go#1963]
### What version are you using?

Horizon 0.23.1

### What did you do?

I checked response times pre/post upgrade from 0.22.2 to 0.23.1. I was surprised that response times for `/order_book` are actually higher even though 0.23.0 started using in-memory offers graph. Count of responses with `duration > 0.1`:

![Screenshot 2019-11-20 at 23 51 31](https://user-images.githubusercontent.com/464938/69285415-255b8480-0bf1-11ea-90ce-0701198e9df5.png)

Not super urgent because p99 of `duration` for this route is actually below 0.10-0.13 but it's worth checking why it's slow (deploy around 22:10):

![Screenshot 2019-11-20 at 23 58 04](https://user-images.githubusercontent.com/464938/69285621-9b5feb80-0bf1-11ea-8cca-2f9162283df3.png)

### What did you expect to see?

Smaller `duration` of `/order_book` responses.

### What did you see instead?

Higher `duration` of `/order_book` responses.

**Comment by bartekn** [comment]
Thanks for checking this @tamirms! Super interesting. Too bad it's hard to load the old data and check the metrics once again (maybe I did something wrong with my query). I zoomed out to check 7d charts and found that in-memory `/order_book` p99 and avg were slightly higher from time to time (still lower than DB version) so I started wondering that maybe it has something to do with a type of queries (`limit` or market params actually affect response time):

<img width="807" alt="Screenshot 2020-06-11 at 00 44 51" src="https://user-images.githubusercontent.com/464938/84326231-c5243400-ab7c-11ea-960d-a679c23cf230.png">

Anyway, I think the response times are great even with a DB version so maybe we shouldn't revert it. Could be interesting to compare CPU profile of 0.23.1 and 1.3.0. Maybe it will give us some useful hints connected to graph optimization.

**Comment by ire-and-curses** [comment]
I'm against reverting because this is big for us because i) it means we (or anyone) can scale request-serving Horizons horizontally through read-only replicas and ii) it removes the need for request-serving Horizons to ingest, breaking the N-N mapping between Horizon and core. This is a big deal for fast txmeta scalability.

**Comment by bartekn** [comment]
@ire-and-curses sorry, I wasn't clear. I meant reverting to the code that's using order-book graph as a data source instead of a DB (reading part), not reverting ingestion on front-end nodes (writing part). Once graph is updated by `OrderBookStream` it can be used by any Horizon component (in the previous code `/order_book` was using the same graph as `/paths`).

**Comment by tamirms** [comment]
Sorry I should have clarified that, since Horizon 1.4.0, the order book graph is populated by polling the Horizon DB instead of via distributed ingestion https://github.com/stellar/go/pull/2630. This means the order book graph is available on all the request serving Horizon nodes. So we could go back to querying the order book graph instead of the DB to serve /order_book responses. Doing so would **not** require the request serving Horizon nodes to participate in ingestion again.

The reason why I implemented orderbook queries using the Horizon DB in https://github.com/stellar/go/pull/2617 is because I thought reducing our dependency on the in memory order book graph would make it easier to remove the frontend Horizon nodes from participating in distributed ingestion. While implementing https://github.com/stellar/go/pull/2630 I realized that we could still have an in memory order book graph without having to force the frontend nodes to participate in ingestion.

If we went back to querying the in memory order book graph I would want to check that the extra queries on the order book graph would not have a negative impact in terms of contention. A read write lock ensures that the order book graph can be updated in a thread safe manner while other go routines are trying to read from the order book graph.

If we ever want to join information from the order book db query with data found in other tables that will be easier than combining in memory order book graph queries with db queries.
